src/shogun/base/SGObject.h

```diff
@@ -1306,13 +1315,30 @@ void CSGObject::observe(
 	this->observe(obs);
 }
 
-template <class T>
+template<class T,
+		typename std::enable_if_t<
+				!is_sg_vector<T>::value &&
+				!is_sg_matrix<T>::value>*>
 void CSGObject::observe(const int64_t step, const std::string& name) const
 {
 	auto param = this->get_parameter(BaseTag(name));
 	auto obs = some<ObservedValueTemplated<T>>(
 		step, name, any_cast<T>(param.get_value()), param.get_properties());
 	this->observe(obs);
 }
+
+template<class T,
+		typename std::enable_if_t<
+				is_sg_vector<T>::value ||
```

Comment:
what about std::vector?

Reply:
btw this is very similar to the problem we have in `any.h` with `clone` and `equals`. There we do the following: we check (in order) whether a type has `->clone`, then `.clone`, and if not we use the copy ctor. We also cover cases such as raw arrays and `std::vector`. I think an approach like this would be better here as well, as otherwise we will start patching and patching the template signature here, causing a change in `SGObject` every time....

Reply:
Do you mean something like this https://github.com/shogun-toolbox/shogun/blob/develop/src/shogun/lib/any.h#L276?

Reply:
> what about std::vector?

`std::vector` has a copy constructor so there is no problem. The issue was with `SGVector` since its copy constructor does not copy also the memory, but it copies only the reference to the memory (hence we have multiple `SGVector`s that point to the same location).

Reply:
@karlnapf have a look at this one: https://github.com/geektoni/shogun/commit/0a9a92e8012fc7352e4d757cb85f0681145854fa. This is basically the `any.h` code, but it should do exactly what we are looking for. Still WIP though. Feel free to criticize it :)

Reply:
Cool! :) That seems ok to me. The only critique I have is now we have the code twice within the codebase.... maybe you can make use of that in `any.h` too and delete it there? Or instead use the implementation from the `any_detail` namespace (since any is included in your file)
If not, also the filename can maybe just be `clone.h`

Reply:
The `any.h` implementation uses some casts to `void*` and it also uses a `void** storage` argument to store the cloned value. These are things which are not exactly needed by our use case, so I thought it would have been better to have a separate `clone` which can be used from Shogun (and that is not `Any` specific).

Reply:
I see, seems reasonable ... :)